python/cudf/cudf/core/series.py

```diff
@@ -1415,31 +1416,56 @@ def as_mask(self):
         """
         return self._column.as_mask()
 
-    def astype(self, dtype, errors="raise", **kwargs):
+    def astype(self, dtype, copy=False, errors="raise", **kwargs):
         """
         Cast the Series to the given dtype
 
         Parameters
         ----------
 
-        dtype : data type
+        dtype : data type, or dict of column name -> data type
+            Use a numpy.dtype or Python type to cast Series object to
+            the same type. Alternatively, use {col: dtype, ...}, where col is a
+            series name and dtype is a numpy.dtype or Python type to cast to.
+        copy : bool, default False
+            Return a deep-copy when ``copy=True``. Note by default
+            ``copy=False`` setting is used and hence changes to
+            values then may propagate to other cudf objects.
+        errors : {'raise', 'ignore', 'warn'}, default 'raise'
+            Control raising of exceptions on invalid data for provided dtype.
+            - ``raise`` : allow exceptions to be raised
+            - ``ignore`` : suppress exceptions. On error return original
+            object.
+            - ``warn`` : prints last exceptions as warnings and
+            return original object.
         **kwargs : extra arguments to pass on to the constructor
 
         Returns
         -------
         out : Series
-            Copy of ``self`` cast to the given dtype. Returns
-            ``self`` if ``dtype`` is the same as ``self.dtype``.
+            Returns ``self.copy(deep=copy)`` if ``dtype`` is the same
+            as ``self.dtype``.
         """
         if errors not in ("ignore", "raise", "warn"):
             raise ValueError("invalid error value specified")
 
+        if is_dict_like(dtype):
+            if len(dtype) > 1 or self.name not in dtype:
+                raise KeyError(
+                    "Only the Series name can be used for "
+                    "the key in Series dtype mappings."
+                )
+            dtype = dtype[self.name]
+
         if pd.api.types.is_dtype_equal(dtype, self.dtype):
-            return self
+            return self.copy(deep=copy)
         try:
+            data = self._column.astype(dtype, **kwargs)
+
             return self._copy_construct(
-                data=self._column.astype(dtype, **kwargs)
+                data=data.copy(deep=copy), index=self.index.copy(deep=copy)
```

Comment:
Can't shallow copy the column here if `copy=False` as that will then break inplace operations down the line.

Reply:
Also index should always be passed by reference.

Reply:
But `copy=False` means a shallow copy?
and when `copy=True` we are doing a shallow copy of constructor but doing a deep-copy of the actual `_column`.

> ...as that will then break inplace operations down the line.

I didn't get this?

Reply:
So they way we handle inplace operations in the case of something like a StringColumn is to actually just swap the Buffers out from underneath a Column object. This way a Column is the unit of data from the perspective of a DataFrame / Series / Index / etc. So if we shallow copy the Column object and then run an inplace operation that swaps the buffers, we end up swapping the buffers against a new Column object as opposed to the one that could be referenced elsewhere.

Reply:
i.e.
```
gds = cudf.Series(['a', 'b', 'c'])
gdf = cudf.DataFrame()
gdf['col'] = gds
gds = gds.astype('str', copy=False)
gdf['col'][0] = 'blah'
print(gds) # should reflect the changes from modifying `gdf['col']`, but wouldn't because we shallow copied the column underneath
```

Reply:
Thanks @kkraus14  for clarifying patiently. 🙏